addons/purchase/purchase.py

```diff
@@ -1238,8 +1238,14 @@ def _run(self, cr, uid, procurement, context=None):
         return super(procurement_order, self)._run(cr, uid, procurement, context=context)
 
     def _check(self, cr, uid, procurement, context=None):
-        if procurement.purchase_line_id and procurement.purchase_line_id.order_id.shipped:  # TOCHECK: does it work for several deliveries?
-            return True
+        if procurement.purchase_line_id:
+            if procurement.purchase_line_id.order_id.shipped:
+                return True
+            elif procurement.move_ids:
+                moves = self.pool.get('stock.move').browse(cr, uid, [x.id for x in procurement.move_ids], context=context)
+                done_or_cancel = all(move.state == 'done' or move.state == 'cancel' for move in moves)
+                all_cancel = all(move.state == 'cancel' for move in moves)
```

Comment:
All done instead of done/cancel would do it also as the user can put the shipped flag later which will put it done also.  But it is debatable.